File: Trees/Python/Design a Number Container System.py
# Problem link: https://leetcode.com/problems/design-a-number-container-system/description/
# Time: O(log n)
# Space: O(n)
from sortedcontainers import SortedSet

class NumberContainers:

    def __init__(self):
        self.index_mapping = {}
        self.mapping = defaultdict(SortedSet)

    def change(self, index: int, number: int) -> None:
        if index in self.index_mapping:
            num = self.index_mapping[index]
            self.mapping[num].remove(index)
        
        self.index_mapping[index] = number
        self.mapping[number].add(index)

    def find(self, number: int) -> int:
        if self.mapping[number]:
            return self.mapping[number][0]
        return -1

File: Trees/Java/Design a Number Container System.java
// Problem link: https://leetcode.com/problems/design-a-number-container-system/description/
// Time: O(log n)
// Space: O(n)
class NumberContainers {

    Map<Integer, Integer> index_mapping;
    Map<Integer, TreeSet<Integer>> mapping;

    public NumberContainers() {
        this.index_mapping = new HashMap<>();
        this.mapping = new HashMap<>();
    }
    
    public void change(int index, int number) {
        if(this.index_mapping.containsKey(index)) {
            int num = this.index_mapping.get(index);
            this.mapping.get(num).remove(index);
            if(this.mapping.get(num).size() == 0) {
                this.mapping.remove(num);
            }
        }

        this.mapping.putIfAbsent(number, new TreeSet<>());
        this.mapping.get(number).add(index);
        this.index_mapping.put(index, number);
    }
    
    public int find(int number) {
        if(this.mapping.containsKey(number)) {
            return this.mapping.get(number).first();
        }
        return -1;
    }
}
